ConnectWise PSA (Independent Publisher)

howellchrisj opened the pull request:
---
### When submitting a connector, please make sure that you follow the requirements below, otherwise your PR might be rejected. We want to make you have a well-built connector, a smooth certification experience, and your users are happy :) 

If this is your first time submitting to GitHub and you need some help, please sign up for this [session](https://forms.office.com/pages/responsepage.aspx?id=KtIy2vgLW0SOgZbwvQuRaXDXyCl9DkBHq4A2OG7uLpdUMTFJWFFGVUxBNUFZQjZWRUdaOE5BMFkwNS4u). 

- [ ] I attest that the connector doesn't exist on the Power Platform today. I've verified by checking the pull requests in GitHub and by searching for the connector on the platform or in the documentation. 
- [ ] I attest that the connector works and I verified by deploying and testing all the operations. 
- [ ] I attest that I have added detailed descriptions for all operations and parameters in the swagger file.
- [ ] I attest that I have added response schemas to my actions, unless the response schema is dynamic. 
- [ ] I validated the swagger file, `apiDefinition.swagger.json`, by running `paconn validate` command.
- [ ] If this is a certified connector, I confirm that `apiProperties.json` has a valid brand color and doesn't use an invalid brand color, `#007ee5` or `#ffffff`. If this is an independent publisher connector, I confirm that I am not submitting a connector icon.

If you are an Independent Publisher, you must also attest to the following to ensure a smooth publishing process:
- [ ] I have named this PR after the pattern of "Connector Name (Independent Publisher)" ex: HubSpot Marketing (Independent Publisher)
- [ ] Within this PR markdown file, I have pasted screenshots that show: 3 unique operations (actions/triggers) working within a Flow. This can be in one flow or part of multiple flows. For each one of those flows, I have pasted in screenshots of the Flow succeeding. 
- [ ] Within this PR markdown file, I have pasted in a screenshot from the Test operations section within the Custom Connector UI.
- [ ] If the connector uses OAuth, I have provided detailed steps on how to create an app in the readme.md.

howellchrisj committed:
Create README.md

howellchrisj committed:
Update README.md

howellchrisj committed:
Screenshot

howellchrisj committed:
ConnectWise API Connector

howellchrisj committed:
Add files via upload

howellchrisj committed:
Update README.md

howellchrisj committed:
Add files via upload

howellchrisj committed:
Delete independent-publisher-connectors/Proposal - ConnectWise (Independent Publisher)/apiDefinition.swagger.json

howellchrisj committed:
Rename ConnectWise-API-Connector.swagger.json to apiDefinition.swagger.json

howellchrisj committed:
Add files via upload

howellchrisj committed:
Delete independent-publisher-connectors/Proposal - ConnectWise (Independent Publisher)/independent-publisher-icon.png

howellchrisj committed:
Delete independent-publisher-connectors/Proposal - ConnectWise (Independent Publisher)/ConnectWise API Connector 1.png

howellchrisj committed:
Delete independent-publisher-connectors/Proposal - ConnectWise (Independent Publisher)/ConnectWise API Connector 2.png

howellchrisj committed:
Add files via upload

howellchrisj committed:
Add files via upload

howellchrisj committed:
Update README.md

howellchrisj committed:
Update apiDefinition.swagger.json

howellchrisj committed:
Update apiDefinition.swagger.json

howellchrisj committed:
Add files via upload

howellchrisj committed:
Update apiDefinition.swagger.json

Added Metadata section

howellchrisj committed:
Add files via upload

howellchrisj committed:
Create README.md

howellchrisj committed:
Add files via upload

howellchrisj committed:
Delete independent-publisher-connectors/Proposal - ConnectWise (Independent Publisher)/apiProperties.json

howellchrisj committed:
Delete independent-publisher-connectors/Proposal - ConnectWise (Independent Publisher)/README.md

howellchrisj committed:
Delete independent-publisher-connectors/Proposal - ConnectWise (Independent Publisher)/apiDefinition.swagger.json

howellchrisj committed:
Delete independent-publisher-connectors/README.md

howellchrisj committed:
Update apiDefinition.swagger.json

howellchrisj committed:
Update apiProperties.json

howellchrisj committed:
Merge branch 'dev' into dev

howellchrisj committed:
Update apiProperties.json

howellchrisj committed:
Update apiProperties.json

Trying to get the Authentication to appear in CW Connector

howellchrisj committed:
Merge branch 'dev' into dev

howellchrisj committed:
Update apiProperties.json

howellchrisj committed:
Update apiDefinition.swagger.json

howellchrisj committed:
Update apiDefinition.swagger.json

Corrected swagger issues

howellchrisj committed:
Update apiProperties.json

howellchrisj committed:
Update apiProperties.json

howellchrisj committed:
Update apiDefinition.swagger.json

howellchrisj committed:
Merge branch 'dev' into dev

vmanoharas reviewed (commented):
Hi @howellchrisj, 

Could you please clarify the below swagger validation error? 

##[error]The 'host' property, if present, must be a valid URI host without a scheme or path. ValuePath: /host

troystaylor commented:
> Hi @howellchrisj,
> 
> Could you please clarify the below swagger validation error?
> 
> ##[error]The 'host' property, if present, must be a valid URI host without a scheme or path. ValuePath: /host

I would recommend setting the host here to something (e.g. microsoft.com) and then using a policy to set it from the connection parameter: https://learn.microsoft.com/en-us/connectors/custom-connectors/policy-templates/dynamichosturl/dynamichosturl

howellchrisj committed:
Update apiProperties.json

howellchrisj committed:
Update apiDefinition.swagger.json

vmanoharas commented:
> > Hi @howellchrisj,
> > Could you please clarify the below swagger validation error?
> > ##[error]The 'host' property, if present, must be a valid URI host without a scheme or path. ValuePath: /host
> 
> I would recommend setting the host here to something (e.g. microsoft.com) and then using a policy to set it from the connection parameter: https://learn.microsoft.com/en-us/connectors/custom-connectors/policy-templates/dynamichosturl/dynamichosturl

Hi @howellchrisj, 

Can you please address comments?

howellchrisj commented:
@vmanoharas , I adjusted the code to have the URL within the json file per the recommendations.  There are two things, one, I changed the Authorization to get an API Key to work and then during that change, I realized that the URL has to be somewhat dynamic as I am using "na", but other countries may need to use "eu" in the URL.   For the Default I put it is as the "na", but have the parameter to allow for the change.

vmanoharas commented:
> @vmanoharas , I adjusted the code to have the URL within the json file per the recommendations. There are two things, one, I changed the Authorization to get an API Key to work and then during that change, I realized that the URL has to be somewhat dynamic as I am using "na", but other countries may need to use "eu" in the URL. For the Default I put it is as the "na", but have the parameter to allow for the change.

Hi @howellchrisj,  we would recommend you to test this functionality thoroughly and confirm with us as we will not test this from our end.  Additionally, please post the screenshots of successful flow creation.

howellchrisj commented:
@vmanoharas , I would love to test this flow.  Can you provide me details on how I am supposed to test the connector without the changes being pushed into production?  Every time I wrote the custom connector and tried to publish, the exact syntax and json wasn't the same as what it needs to be for the publishing updates.

vmanoharas commented:
Hi @howellchrisj, 

We would recommend you to refer other IP PRs as they tested and posted the screenshots before we approve.  Kindly refer the below links. 

https://github.com/microsoft/PowerPlatformConnectors?tab=readme-ov-file#independent-publisher-connectors

https://learn.microsoft.com/en-us/connectors/custom-connectors/certification-submission-ip

https://github.com/microsoft/PowerPlatformConnectors/tree/dev/templates/Independent%20Publisher

https://github.com/microsoft/PowerPlatformConnectors/tree/dev/schemas

howellchrisj committed:
Update apiProperties.json

howellchrisj committed:
Update apiDefinition.swagger.json

howellchrisj commented:
Completed the testing and updating of the .json files and tested via Custom Connector

![image](https://github.com/user-attachments/assets/57708240-4981-4b49-a260-4a48acf45a7b)

howellchrisj committed:
Update apiProperties.json

certificationteam reviewed (approved):
Dear Partner

I hope you are doing well.
Congratulations, your pull request is approved. We will move forward with the certification process. We are preparing your connector for production deployment and will queue up for the next deployment schedule. Once your connector onboards the next deployment schedule, it starts to deploy your connector in our production environments that typically takes 3-4 weeks.
Please let us know if you have any questions.
Thank you very much for working with us.

vmanoharas reviewed (approved):
Dear Partner

I hope you are doing well.
Congratulations, your pull request is approved. We will move forward with the certification process. We are preparing your connector for production deployment and will queue up for the next deployment schedule. Once your connector onboards the next deployment schedule, it starts to deploy your connector in our production environments that typically takes 3-4 weeks.
Please let us know if you have any questions.
Thank you very much for working with us.

vmanoharas commented:
Hello Partner,
I hope you are doing well.
Congratulations, the deployment of your connector and documentation page was already completed, 
Thank you very much for working with us.
